src/Features/Core/Portable/AddImport/SymbolReferenceFinder_PackageAssemblySearch.cs

```diff
@@ -90,22 +86,71 @@ private async Task FindReferenceAssemblyTypeReferencesAsync(
             CancellationToken cancellationToken)
         {
             cancellationToken.ThrowIfCancellationRequested();
-            var results = await _symbolSearchService.FindReferenceAssembliesWithTypeAsync(
-                name, arity, cancellationToken).ConfigureAwait(false);
+            var results = await _symbolSearchService.FindReferenceAssembliesAsync(
+                name, arity, _isWithinImport, cancellationToken).ConfigureAwait(false);
 
             var project = _document.Project;
 
             foreach (var result in results)
             {
                 cancellationToken.ThrowIfCancellationRequested();
+                if (!IncludeResult(nameNode, result.ContainingNamespaceNames))
+                    continue;
+
                 await HandleReferenceAssemblyReferenceAsync(
                     allReferences, nameNode, project,
                     isAttributeSearch, result, weight: allReferences.Count,
                     cancellationToken: cancellationToken).ConfigureAwait(false);
             }
         }
 
-        private async Task FindNugetTypeReferencesAsync(
+        private bool IncludeResult(TSimpleNameSyntax nameNode, ImmutableArray<string> namespaceNames)
+        {
+            // Outside of a using/import we did a search based on a simple leftmost name.  The index only provides
+            // results for namespaces that contain that name.  So we don't need to do anything special here and we can
+            // include that result.
+            if (!_isWithinImport)
+                return true;
+
+            // Inside of a using/import we do a search on the part of the using that doesn't bind (like 'Json' in `using
+            // Newtonsoft.Json;`).  But this may find results in other potential namespaces (like `Goobar.Json`).  So we
+            // need to make sure the namespace the index found matches the full name in the using/import.
```

Comment:
new logic needed for 'add package for namespace' functionality.  previously we only ever searched for a simple name like 'JToken' just seeing it matched a top level type.  Now we're actually searching for results that can match a possibly dotted namespace name (like `Newtonsoft` or `Newtonsoft.Json`.  So we have to do more work validating the result.

Reply:
it happns here as this is the location that has access to syntax/semantics to check the result against.

i could have also grabbed out the names and passed them into the search function.  but this was simple enough to do in this fashion.